conjure-java-core/src/main/java/com/palantir/conjure/java/types/ReturnTypeClassNameVisitor.java

```diff
@@ -63,15 +64,28 @@ public TypeName visitOptional(OptionalType type) {
     @Override
     public TypeName visitPrimitive(PrimitiveType type) {
         if (type.get() == PrimitiveType.Value.BINARY) {
-            return RESPONSE_BODY_TYPE;
+            return binaryClassName;
         } else {
             return delegate.visitPrimitive(type);
         }
     }
 
     @Override
     public TypeName visitReference(com.palantir.conjure.spec.TypeName type) {
-        return BinaryReturnTypeResolver.resolveReturnReferenceType(types, type, RESPONSE_BODY_TYPE);
+        if (!types.containsKey(type)) {
+            throw new IllegalStateException("Unknown LocalReferenceType type: " + type);
+        }
+
+        TypeDefinition def = types.get(type);
+        if (def.accept(TypeDefinitionVisitor.IS_ALIAS)) {
+            Type aliasType = def.accept(TypeDefinitionVisitor.ALIAS).getAlias();
+            TypeName aliasTypeName = aliasType.accept(this);
```

Comment:
Long term I'd like to completely get rid of this class - it's kinda doing some 'dealiasing' here, but it's pretty convoluted with the binary specialization we have going on above.